File: python/producer/src/producer.py
import json

import pika
from pika.adapters import BlockingConnection


class Producer:
    def __init__(self, queue: str):
        self.queue = queue
        self.connection = self.connect()
        self.channel = self.connection.channel()
        self.channel.queue_declare(queue=queue, durable=True)

    def connect(self) -> BlockingConnection:
        connection_parameters = pika.ConnectionParameters("localhost", 5672)
        return pika.BlockingConnection(connection_parameters)

    def publish(self, message: dict):
        self.channel.basic_publish(
            exchange="", routing_key=self.queue, body=json.dumps(message)
        )
        print(f" [x] Sent '{message}'")

    def close(self):
        self.connection.close()


class ProducerFactory:
    @staticmethod
    def create_producer(queue: str) -> Producer:
        return Producer(queue)

File: nodejs/producer/src/producer.ts
import amqp, { Channel } from "amqplib";

export class Producer {
  private channel: Channel | undefined;
  private queue: string | undefined;

  async connect(queue: string): Promise<void> {
    try {
      const connection = await amqp.connect("amqp://localhost:5672");

      this.queue = queue;
      this.channel = await connection.createChannel();
      await this.channel.assertQueue(this.queue);
    } catch (e) {
      console.error(e);
    }
  }

  async produce(message: object): Promise<void> {
    if (!this.channel || !this.queue) {
      throw new Error("Producer not connected.");
    }

    this.channel.sendToQueue(this.queue, Buffer.from(JSON.stringify(message)));
  }
}

export class ProducerFactory {
  static async create(queue: string): Promise<Producer> {
    const producer = new Producer();
    await producer.connect(queue);

    return producer;
  }
}
